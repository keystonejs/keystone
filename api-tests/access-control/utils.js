const { setupServer } = require('@keystonejs/test-utils');
const { Text } = require('@keystonejs/fields');
const { objMerge } = require('@keystonejs/utils');

const FAKE_ID = { mongoose: '5b3eabd9e9f2e3e4866742ea', knex: 137 };
const FAKE_ID_2 = { mongoose: '5b3eabd9e9f2e3e4866742eb', knex: 138 };

const yesNo = truthy => (truthy ? 'Yes' : 'No');

const getPrefix = access => {
  // prettier-ignore
  let prefix = `${yesNo(access.create)}Create${yesNo(access.read)}Read${yesNo(access.update)}Update${yesNo(access.auth)}Auth`;
  if (Object.prototype.hasOwnProperty.call(access, 'delete')) {
    prefix = `${prefix}${yesNo(access.delete)}Delete`;
  }
  return prefix;
};

const getStaticListName = access => `${getPrefix(access)}StaticList`;
const getImperativeListName = access => `${getPrefix(access)}ImperativeList`;
const getDeclarativeListName = access => `${getPrefix(access)}DeclarativeList`;

/* Generated with:
  const result = [];
  const options = ['create', 'read', 'update', 'delete', 'auth'];
  // All possible combinations are contained in the set 0..2^n-1
  for(let flags = 0; flags < Math.pow(2, options.length); flags++) {
    // Generate an object of true/false values for the particular combination
    result.push(options.reduce((memo, option, index) => ({
      ...memo,
      // Use a bit mask to see if that bit is set
      [option]: !!(flags & (1 << index)),
    }), {}));
  }
  */
const listAccessVariations = [
  { create: false, read: false, update: false, delete: false, auth: true },
  { create: true, read: false, update: false, delete: false, auth: true },
  { create: false, read: true, update: false, delete: false, auth: true },
  { create: true, read: true, update: false, delete: false, auth: true },
  { create: false, read: false, update: true, delete: false, auth: true },
  { create: true, read: false, update: true, delete: false, auth: true },
  { create: false, read: true, update: true, delete: false, auth: true },
  { create: true, read: true, update: true, delete: false, auth: true },
  { create: false, read: false, update: false, delete: true, auth: true },
  { create: true, read: false, update: false, delete: true, auth: true },
  { create: false, read: true, update: false, delete: true, auth: true },
  { create: true, read: true, update: false, delete: true, auth: true },
  { create: false, read: false, update: true, delete: true, auth: true },
  { create: true, read: false, update: true, delete: true, auth: true },
  { create: false, read: true, update: true, delete: true, auth: true },
  { create: true, read: true, update: true, delete: true, auth: true },
  { create: false, read: false, update: false, delete: false, auth: false },
  { create: true, read: false, update: false, delete: false, auth: false },
  { create: false, read: true, update: false, delete: false, auth: false },
  { create: true, read: true, update: false, delete: false, auth: false },
  { create: false, read: false, update: true, delete: false, auth: false },
  { create: true, read: false, update: true, delete: false, auth: false },
  { create: false, read: true, update: true, delete: false, auth: false },
  { create: true, read: true, update: true, delete: false, auth: false },
  { create: false, read: false, update: false, delete: true, auth: false },
  { create: true, read: false, update: false, delete: true, auth: false },
  { create: false, read: true, update: false, delete: true, auth: false },
  { create: true, read: true, update: false, delete: true, auth: false },
  { create: false, read: false, update: true, delete: true, auth: false },
  { create: true, read: false, update: true, delete: true, auth: false },
  { create: false, read: true, update: true, delete: true, auth: false },
  { create: true, read: true, update: true, delete: true, auth: false },
];

const fieldMatrix = [
  { create: false, read: false, update: false },
  { create: true, read: false, update: false },
  { create: false, read: true, update: false },
  { create: true, read: true, update: false },
  { create: false, read: false, update: true },
  { create: true, read: false, update: true },
  { create: false, read: true, update: true },
  { create: true, read: true, update: true },
];
const getFieldName = access => getPrefix(access);

const nameFn = { imperative: getImperativeListName, declarative: getDeclarativeListName };

const createFieldStatic = fieldAccess => ({
  [getFieldName(fieldAccess)]: {
    type: Text,
    access: fieldAccess,
  },
});
const createFieldImperative = fieldAccess => ({
  [getFieldName(fieldAccess)]: {
    type: Text,
    access: {
      create: () => fieldAccess.create,
      read: () => fieldAccess.read,
      update: () => fieldAccess.update,
    },
  },
});
function setupKeystone(adapterName) {
  return setupServer({
    adapterName,
    createLists: keystone => {
      keystone.createList('User', { fields: { name: { type: Text } } });

      listAccessVariations.forEach(access => {
        keystone.createList(getStaticListName(access), {
          fields: {
            name: { type: Text },
            ...objMerge(fieldMatrix.map(variation => createFieldStatic(variation))),
          },
          access,
        });
        keystone.createList(getImperativeListName(access), {
          fields: {
            name: { type: Text },
            ...objMerge(fieldMatrix.map(variation => createFieldImperative(variation))),
          },
          access: {
            create: () => access.create,
            read: () => access.read,
            update: () => access.update,
            delete: () => access.delete,
            auth: () => access.auth,
          },
        });
        keystone.createList(getDeclarativeListName(access), {
          fields: { name: { type: Text } },
          access: {
            create: access.create,
            // arbitrarily restrict the data to a single item (see data.js)
            read: () => access.read && { name_starts_with: 'Hello' },
            update: () => access.update && { name_starts_with: 'Hello' },
            delete: () => access.delete && { name_starts_with: 'Hello' },
            auth: access.auth,
          },
        });
      });
    },
  });
}
module.exports = {
  FAKE_ID,
  FAKE_ID_2,
  getStaticListName,
  getImperativeListName,
  getDeclarativeListName,
  listAccessVariations,
  fieldMatrix,
  nameFn,
  setupKeystone,
  getFieldName,
};
