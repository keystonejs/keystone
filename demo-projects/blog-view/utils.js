const terminalLink = require('terminal-link');
const ciInfo = require('ci-info');
const chalk = require('chalk');

const ttyLink = (text, path, port) => {
  if (ciInfo.isCI) {
    return;
  }
  const url = `http://localhost:${port}${path}`;
  const link = terminalLink(url, url, { fallback: () => url });
  console.log(`ðŸ”— ${chalk.green(text)}\t${link}`);
};

function extractAppMeta(apps) {
  let adminPath;
  let graphiqlPath;
  let apiPath;

  apps.forEach(app => {
    switch (app.constructor.name) {
      case 'AdminUIApp': {
        adminPath = app.adminPath;
        break;
      }
      case 'GraphQLApp': {
        apiPath = app._apiPath;
        graphiqlPath = app._graphiqlPath;
        break;
      }
    }
  });

  return {
    adminPath,
    graphiqlPath,
    apiPath,
  };
}

function logAdminRoutes(apps, port) {
  const { adminPath, graphiqlPath, apiPath } = extractAppMeta(apps);
  /* eslint-disable no-unused-expressions */
  adminPath && ttyLink('Keystone Admin UI:', adminPath, port);
  graphiqlPath && ttyLink('GraphQL Playground:', graphiqlPath, port);
  apiPath && ttyLink('GraphQL API:\t', apiPath, port);
  /* eslint-enable no-unused-expressions */
}

module.exports = {
  logAdminRoutes,
};
