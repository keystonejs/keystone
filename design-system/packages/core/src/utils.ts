import {
  ComponentPropsWithoutRef,
  ElementType,
  forwardRef,
  ReactElement,
  ReactNode,
  Ref,
  useEffect,
  useLayoutEffect,
  useState,
} from 'react';
import { createPortal } from 'react-dom';

/*
  @johannes' one weird trick for fixing TypeScript autocomplete
*/
export function identityType<T>() {
  function inner<U extends T>(u: U): U {
    return u;
  }
  return inner;
}

/*
  Logs a warning to the console when the condition is true, only in dev
*/
export const devWarning = (condition: boolean, message: string) => {
  if (process.env.NODE_ENV !== 'production') {
    if (condition) {
      console.error(message);
    }
  }
};

/*
  forwardRefWithAs lets us forward refs while keeping the correct component type,
  which can be specified by the `as` prop.
*/

type ElementTagNameMap = HTMLElementTagNameMap &
  Pick<SVGElementTagNameMap, Exclude<keyof SVGElementTagNameMap, keyof HTMLElementTagNameMap>>;

type AsProp<Comp extends ElementType, Props> = {
  as?: Comp;
  ref?: Ref<
    Comp extends keyof ElementTagNameMap
      ? ElementTagNameMap[Comp]
      : Comp extends new (...args: any) => any
      ? InstanceType<Comp>
      : undefined
  >;
} & Omit<ComponentPropsWithoutRef<Comp>, 'as' | keyof Props>;

type CompWithAsProp<Props, DefaultElementType extends ElementType> = <
  Comp extends ElementType = DefaultElementType
>(
  props: AsProp<Comp, Props> & Props
) => ReactElement;

export const forwardRefWithAs = <DefaultElementType extends ElementType, BaseProps>(
  render: (
    props: BaseProps & { as?: ElementType },
    ref: React.Ref<any>
  ) => Exclude<ReactNode, undefined>
): CompWithAsProp<BaseProps, DefaultElementType> => {
  // @ts-ignore
  return forwardRef(render);
};

/*
  A helper for making valid IDs from a set of inputs
*/
export function makeId(...args: (string | number | null | undefined)[]) {
  return args.filter(val => val != null).join('--');
}

/*
  A helper for handling string OR array values e.g.

  <Component size="small" />
  VS
  <Component size={['small', 'large']} />
*/
export const mapResponsiveProp = <
  Map extends Record<string, string | number>,
  Keys extends keyof Map
>(
  value: Keys | readonly (Keys | null)[],
  valueMap: Map
) => {
  if (Array.isArray(value)) {
    return value.map(k => (k == null ? null : valueMap[k]));
  }
  // @ts-ignore
  return valueMap[value];
};

/**
 * Utils below are ported with thanks from @reach-ui
 * Copyright (c) 2018-present, React Training LLC
 */

// Autogenerate IDs to facilitate WAI-ARIA and server rendering. For reasoning, see
// https://github.com/reach/reach-ui/blob/develop/packages/auto-id/src/index.tsx
let serverHandoffComplete = false;
let id = 0;
const genId = () => ++id;

export const useId = (idFromProps?: string | null) => {
  const initialId = idFromProps || (serverHandoffComplete ? genId() : null);

  const [id, setId] = useState(initialId);

  useSafeLayoutEffect(() => {
    if (id === null) {
      setId(genId());
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (serverHandoffComplete === false) {
      serverHandoffComplete = true;
    }
  }, []);
  return id != null ? String(id) : undefined;
};

// Works around useLayoutEffect throwing a warning when used in SSR
export const useSafeLayoutEffect = typeof window === 'undefined' ? () => {} : useLayoutEffect;

type Props = {
  children: ReactElement;
};

export const Portal = ({ children }: Props) => {
  if (typeof document === 'undefined') {
    return null;
  }

  return createPortal(children, document.body);
};
