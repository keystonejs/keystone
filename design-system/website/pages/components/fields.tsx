/* @jsx jsx */

import { ReactNode, useState } from 'react';
import { jsx } from '@emotion/core';
import { useTheme } from '@keystone-ui/core';
import { Checkbox, Radio, TextArea, TextInput, Switch } from '@keystone-ui/fields';
import { SegmentedControl } from '@keystone-ui/segmented-control';

import { Page } from '../../components/Page';

const FieldWrapper = ({ children }: { children: ReactNode }) => {
  const { spacing } = useTheme();
  return (
    <div
      css={{
        marginTop: spacing.medium,
        marginBottom: spacing.medium,
      }}
    >
      {children}
    </div>
  );
};

const StatefulSwitch = ({ children, ...props }: { children: ReactNode }) => {
  const [checked, setChecked] = useState(false);
  return (
    <Switch checked={checked} onClick={() => setChecked(!checked)} {...props}>
      {children}
    </Switch>
  );
};

export default function FieldsPage() {
  const { spacing } = useTheme();
  return (
    <Page>
      <h1>Form Fields</h1>
      <h2>Text Inputs</h2>
      <FieldWrapper>
        <TextInput placeholder="Small" size="small" />
      </FieldWrapper>
      <FieldWrapper>
        <TextInput placeholder="Medium" size="medium" />
      </FieldWrapper>
      <FieldWrapper>
        <TextInput placeholder="Large" size="large" />
      </FieldWrapper>
      <h3>Disabled</h3>
      <FieldWrapper>
        <TextInput value="This one is disabled" size="medium" disabled />
      </FieldWrapper>
      <h2>TextAreas</h2>
      <FieldWrapper>
        <TextArea />
      </FieldWrapper>
      <h2>Segmented Controls</h2>
      <FieldWrapper>
        <SegmentedControl animate segments={['one', 'two', 'three']} />
      </FieldWrapper>
      <h2>Checkboxes</h2>
      <FieldWrapper>
        <Checkbox css={{ marginRight: spacing.medium }}>Check Me</Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} checked readOnly>
          Always Checked
        </Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} disabled readOnly>
          Disabled Option
        </Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} disabled checked readOnly>
          Checked + Disabled
        </Checkbox>
      </FieldWrapper>
      <h3>Sizes</h3>
      <FieldWrapper>
        <Checkbox css={{ marginRight: spacing.medium }} size="small">
          Small
        </Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} size="medium">
          Medium
        </Checkbox>
        <Checkbox css={{ marginRight: spacing.medium }} size="large">
          Large
        </Checkbox>
      </FieldWrapper>
      <h2>Radio Buttons</h2>
      <FieldWrapper>
        <Radio name="radio-set" css={{ marginRight: spacing.medium }}>
          Check Me
        </Radio>
        <Radio name="radio-set" css={{ marginRight: spacing.medium }}>
          Or Me
        </Radio>
        <Radio css={{ marginRight: spacing.medium }} checked readOnly>
          Always Checked
        </Radio>
        <Radio css={{ marginRight: spacing.medium }} disabled readOnly>
          Disabled
        </Radio>
        <Radio css={{ marginRight: spacing.medium }} disabled checked readOnly>
          Checked + Disabled
        </Radio>
      </FieldWrapper>
      <h3>Sizes</h3>
      <FieldWrapper>
        <Radio name="radio-size-set" css={{ marginRight: spacing.medium }} size="small">
          Small
        </Radio>
        <Radio name="radio-size-set" css={{ marginRight: spacing.medium }} size="medium">
          Medium
        </Radio>
        <Radio name="radio-size-set" css={{ marginRight: spacing.medium }} size="large">
          Large
        </Radio>
      </FieldWrapper>
      <h2>Switches</h2>
      <FieldWrapper>
        <StatefulSwitch css={{ marginRight: spacing.medium }}>Switch Me</StatefulSwitch>
        <Switch css={{ marginRight: spacing.medium }} checked>
          Always On
        </Switch>
        <Switch css={{ marginRight: spacing.medium }} checked disabled>
          On + Disabled
        </Switch>
      </FieldWrapper>
    </Page>
  );
}
