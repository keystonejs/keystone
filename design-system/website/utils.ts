export const toLabel = (str: string) => str.substr(0, 1).toUpperCase() + str.substr(1);

const VOWELS = ['a', 'e', 'i', 'o', 'u'];
export const aAn = (before: string) => `a${VOWELS.includes(before.charAt(0)) ? 'n' : ''}`;
