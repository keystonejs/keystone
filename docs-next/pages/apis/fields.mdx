import { Markdown } from '../../components/Page';

# Fields API

The `fields` option of a [list configuration](./schema) defines the names, types, and configuration of the fields in the list.
This configuration option takes an object with field names as keys, and configured field types as values.

This document covers the different field types which are available and the configuration options they accept.
To see how to access fields in the GraphQL API please see the [GraphQL API](./graphql) docs.

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import {
  // Scalar types
  checkbox,
  integer,
  password,
  select,
  text,
  timestamp,

  // Relationship type
  relationship,

  // Index types
  autoincrement,
  mongoId,

  // Virtual type
  virtual,
} from '@keystone-next/fields';

// Complex types
import { document } from '@keystone-next/fields-document';
import { cloudinary } from '@keystone-next/cloudinary';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: text({
          /* ... */
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

## Common configuration

All field types accept a common set of configuration options.
All options are optional.

Options:

- `access`: Defines the [Access Control](../guides/access-control) rules for the field.
  See the [Access Control API](./access-control) for full details on the available access control options.
- `hooks`: The `hooks` option defines [hook](../guides/hooks) functions for the field.
  Hooks allow you to execute code at different stages of the mutation lifecycle.
  See the [Hooks API](./hooks) for full details on the available hook options.
- `label`: The label displayed for this field in the Admin UI. Defaults to a human readable version of the field name.
- `ui`: Controls how the field is displayed in the Admin UI.
  - `views`: (coming soon)
  - `description`: (coming soon)
  - `createView`: (coming soon)
  - `listView`: (coming soon)
  - `itemView`: (coming soon)

```typescript
export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: text({
          access: {
            /* ... */
          },
          hooks: {
            /* ... */
          },
          label: '...',
          ui: {
            views: '...',
            description: '...',
            createView: {
              fieldMode: ({ session }) => 'edit',
            },
            listView: {
              fieldMode: ({ session }) => 'read',
            },
            itemView: {
              fieldMode: ({ session, item }) => 'read',
            },
          },
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

## Scalar types

### checkbox

A `checkbox` field represents a boolean (`true`/`false`) value.

Options:

- `defaultValue` (default: `undefined`): Can be either a boolean value or an async function which takes an argument `({ context, originalItem })` and returns a boolean value.
  This value will be used for the field when creating items if no explicit value is set.
  `context` is a [`KeystoneContext`](./context) object.
  `originalItem` is an object containing the data passed in to the `create` mutation.
- `isRequired` (default: `false`): If `true` then this field can never be set to `null`.

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { checkbox } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: checkbox({
          defaultValue: true,
          isRequired: true,
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

### integer

An `integer` field represents an integer value.

Options:

- `defaultValue` (default: `undefined`): Can be either an integer value or an async function which takes an argument `({ context, originalItem })` and returns an integer value.
  This value will be used for the field when creating items if no explicit value is set.
  `context` is a [`KeystoneContext`](./context) object.
  `originalItem` is an object containing the data passed in to the `create` mutation.
- `isRequired` (default: `false`): If `true` then this field can never be set to `null`.
- `isIndexed` (`knex` and `mongoose` adapters only. Default: `false`): If `true` a database level index will be applied to this field, which can make searching faster.
- `isUnique` (default: `false`): If `true` then all values of this field must be unique.

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { integer } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: integer({
          defaultValue: 0,
          isRequired: true,
          isIndexed: true,
          isUnique: true,
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

### password

A `password` field represents an encrypted password value.

Options:

- `isRequired` (default: `false`): If `true` then this field can never be set to `null`.
- `minLength` (default: `8`): The minimum length password accepted.
- `bcrypt` (default: `require('bcryptjs')`): A module which implements the same interface as the [`bcryptjs`](https://www.npmjs.com/package/bcryptjs) package, such as the native [`bcrypt`](https://www.npmjs.com/package/bcrypt) package.
  This module will be used for all encryption routines in the `password` field.

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { password } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: password({
          isRequired: true,
          minLength: 10,
          bcrypt: require('bcrypt'),
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

### select

A `select` field represents the selection of one of fixed set of values.
Values can be either strings, integers, or enum values, as determined by the `dataType` option.
This will determine their GraphQL data type, as well as their database storage type.

Options:

- `dataType` (default: `'string'`): Sets the type of the values of this field.
  Must be one of `['string', 'enum', 'integer']`.
- `options`: An array of `{ label, value }`.
  `label` is a string to be displayed in the Admin UI.
  `value` is either a `string` (for `{ dataType: 'string' }` or `{ dataType: 'enum' }`), or a `number` (for `{ dataType: 'integer' }`).
  The `value` will be used in the GraphQL API and stored in the database.
- `defaultValue`: (default: `undefined`): Can be either a string/integer value or an async function which takes an argument `({ context, originalItem })` and returns a string/integer value.
  This value will be used for the field when creating items if no explicit value is set, and must be one of the values defined in `options`.
  `context` is a [`KeystoneContext`](./context) object.
  `originalItem` is an object containing the data passed in to the `create` mutation.
- `isRequired` (default: `false`): If `true` then this field can never be set to `null`.
- `isIndexed` (`knex` and `mongoose` adapters only. Default: `false`): If `true` a database level index will be applied to this field, which can make searching faster.
- `isUnique` (default: `false`): If `true` then all values of this field must be unique.
- `ui` (default: `{ displayMode: 'select' }`): Configures the display mode of the field in the Admin UI.
  Can be one of `['select', 'segmented-control']`.

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { select } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: select({
          dataType: 'enum',
          options: [
            { label: '...', value: '...' },
            /* ... */
          ],
          defaultValue: '...',
          isRequired: true,
          isIndexed: true,
          isUnique: true,
          ui: { displayMode: 'select' },
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

### text

A `text` field represents a string value.

Options:

- `defaultValue` (default: `undefined`): Can be either a string value or an async function which takes an argument `({ context, originalItem })` and returns a string value.
  This value will be used for the field when creating items if no explicit value is set.
  `context` is a [`KeystoneContext`](./context) object.
  `originalItem` is an object containing the data passed in to the `create` mutation.
- `isRequired` (default: `false`): If `true` then this field can never be set to `null`.
- `isIndexed` (`knex` and `mongoose` adapters only. Default: `false`): If `true` a database level index will be applied to this field, which can make searching faster.
- `isUnique` (default: `false`): If `true` then all values of this field must be unique.
- `ui` (default: `{ displayMode: 'input' }`): Configures the display mode of the field in the Admin UI.
  Can be one of `['input', 'textarea']`.

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { text } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: text({
          defaultValue: '...',
          isRequired: true,
          isIndexed: true,
          isUnique: true,
          ui: { displayMode: 'textarea' },
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

### timestamp

A `timestamp` field represents a time value.

Options:

- `defaultValue` (default: `undefined`): Can be either a string value or an async function which takes an argument `({ context, originalItem })` and returns a string value.
  The string value must be an ISO8601 formatted timestamp string, e.g. `'1970-01-01T00:00:00.000Z'`.
  This value will be used for the field when creating items if no explicit value is set.
  `context` is a [`KeystoneContext`](./context) object.
  `originalItem` is an object containing the data passed in to the `create` mutation.
- `isRequired` (default: `false`): If `true` then this field can never be set to `null`.
- `isIndexed` (`knex` and `mongoose` adapters only. Default: `false`): If `true` a database level index will be applied to this field, which can make searching faster.
- `isUnique` (default: `false`): If `true` then all values of this field must be unique.

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { timestamp } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: timestamp({
          defaultValue: '1970-01-01T00:00:00.000Z',
          isRequired: true,
          isIndexed: true,
          isUnique: true,
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

## Relationship type

### relationship

A `relationship` field represents a relationship between two lists.

(coming soon)

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { relationship } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: relationship({
          ref: '...',
          many: false,
          ui: { /* ... */ }.
          defaultValue: '...',
          isIndexed: '...',
          isUnique: '...',
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

## Index types

### autoincrement

(coming soon)

Options:

- `defaultValue`
- `isRequired`
- `isIndexed` (`knex` and `mongoose` adapters only)
- `isUnique`

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { autoincrement } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: autoincrement({
          defaultValue: 0,
          isRequired: true,
          isIndexed: true,
          isUnique: true,
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

### mongoId

(coming soon)

Options:

- `defaultValue`
- `isRequired`
- `isIndexed` (`knex` and `mongoose` adapters only)
- `isUnique`

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { mongoId } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: mongoId({
          defaultValue: '...',
          isRequired: true,
          isIndexed: true,
          isUnique: true,
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

## Virtual type

### virtual

A `virtual` field represents a value which is computed a read time, rather than stored in the database.

(coming soon)

Options:

- `resolver` (required):
- `graphQLReturnType` (required):
- `graphQLReturnFragment` (default: `undefined` ):
- `extendGraphQLTypes` (default: `[]`):
- `args` (default: `[]`):

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { virtual } from '@keystone-next/fields';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: virtual({
          resolver: (item, args, context, info) => `/* ... */`,
          graphQLReturnType: '...',
          graphQLReturnFragment: '...',
          extendGraphQLTypes: ['...'],
          args: [
            { name: '...', type: 'String' },
            /* ... */
          ],
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

## Complex types

### document

(coming soon)

Options:

- `relationships`
- `componentBlocks`
- `formatting`
- `links`
- `dividers`
- `layouts`

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { document } from '@keystone-next/fields-document';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: document({
          relationships: { /* ... */ };
          componentBlocks: {
            block: { /* ... */ },
            /* ... */
          };
          formatting: { /* ... */ };
          links: true,
          dividers: true,
          layouts: [/* ... */],
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

### cloudinary

(coming soon)

- `isRequired` (default: `false`): If `true` then this field can never be set to `null`.
- `cloudinary`: Configuration for the connected Cloudinary account.
  - `cloudName`
  - `apiKey`
  - `apiSecret`
  - `folder`

```typescript
import { config, createSchema, list } from '@keystone-next/keystone/schema';
import { cloudinary } from '@keystone-next/cloudinary';

export default config({
  lists: createSchema({
    ListName: list({
      fields: {
        fieldName: cloudinary({
          isRequired: true,
          cloudinary: {
            cloudName: process.env.CLOUDINARY_CLOUD_NAME,
            apiKey: process.env.CLOUDINARY_API_KEY;
            apiSecret: process.env.CLOUDINARY_API_SECRET,
            folder: process.env.CLOUDINARY_API_FOLDER,
          },
        }),
        /* ... */
      },
    }),
    /* ... */
  }),
  /* ... */
});
```

export default ({ children }) => <Markdown>{children}</Markdown>;
