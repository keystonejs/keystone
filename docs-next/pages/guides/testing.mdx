import { Markdown } from '../../components/Page';

# Testing

(coming soon)

export default ({ children }) => <Markdown>{children}</Markdown>;
