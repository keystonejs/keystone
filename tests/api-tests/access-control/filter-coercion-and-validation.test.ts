import { text } from '@keystone-6/core/fields';
import { list } from '@keystone-6/core';
import { setupTestRunner } from '@keystone-6/core/testing';
import { apiTestConfig, expectExtensionError } from '../utils';

const runner = setupTestRunner({
  config: apiTestConfig({
    lists: {
      BadAccess: list({
        fields: { name: text() },
        access: {
          filter: {
            query: () => {
              return {
                name: 'blah',
              };
            },
          },
        },
      }),
      Coercion: list({
        fields: { name: text() },
        access: {
          filter: {
            query: () => {
              return { NOT: { name: { equals: 'blah' } } };
            },
          },
        },
      }),
    },
  }),
});

describe('Access control - Filter', () => {
  test(
    'findMany - Bad function return value',
    runner(async ({ context }) => {
      // Valid name
      const { data, errors } = await context.graphql.raw({
        query: `query { badAccesses { id } }`,
      });

      // Returns null and throws an error
      expect(data).toEqual({ badAccesses: null });
      const message =
        'Variable "$where" got invalid value "blah" at "where.name"; Expected type "StringFilter" to be an object.';
      expectExtensionError('dev', false, false, errors, 'Access control', [
        {
          path: ['badAccesses'],
          messages: [`BadAccess.access.filter.query: ${message}`],
          debug: [
            {
              message,
              stacktrace: expect.stringMatching(
                new RegExp(`GraphQLError: ${message.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}`)
              ),
            },
          ],
        },
      ]);
    })
  );
  test(
    'findMany - Coercion',
    runner(async ({ context }) => {
      await context.query.Coercion.createMany({ data: [{ name: 'something' }, { name: 'blah' }] });
      expect(await context.query.Coercion.findMany({ query: 'name' })).toEqual([
        { name: 'something' },
      ]);
    })
  );
});
