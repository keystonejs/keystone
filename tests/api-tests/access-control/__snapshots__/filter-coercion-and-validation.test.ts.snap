// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Access control - Filter findMany - Bad function return value 1`] = `
[
  [GraphQLError: An error occurred while running "Access control".
  - BadAccess.access.filter.query: Variable "$where" got invalid value "blah" at "where.name"; Expected type "StringFilter" to be an object.],
]
`;
